File: Data Structures & Algorithms/trapping-rain-water/submission-5.java
class Solution {
    public int trap(int[] height) {
        int res = 0;
        Deque<Integer> st = new ArrayDeque<>();
        for (int i = 0; i < height.length; i++) {
            int h = height[i];
            while (!st.isEmpty() && height[st.peek()] <= h) {
                int bottomH = height[st.pop()];
                if (st.isEmpty()) {
                    break;
                }
                int left = st.peek();
                int dh = Math.min(height[left], height[i]) - bottomH; // 面积的高
                res += dh * (i - left - 1);
            }
            st.push(i);
        }
        return res;
    }
}

File: Data Structures & Algorithms/daily-temperatures/submission-5.py
class Solution:
    def dailyTemperatures(self, temps: List[int]) -> List[int]:
        n = len(temps)
        res = [0] * n
        st = []
        for i in range(n- 1, -1, -1):
            t = temps[i]
            while st and t >= temps[st[-1]]:
                st.pop() 
            if st:
                res[i] = st[-1] - i
            st.append(i)
        return res